### ISSUE TITLE
Clarify behavior of `trim_trailing_whitespace` on existing lines

### ISSUE THREAD

**Issue by steverice** [editorconfig/editorconfig#208]
I've combed through the documentation and can't find anything that explicitly answers this question:
If I set `trim_trailing_whitespace` to `true`, does that mean trailing whitespace is trimmed from *all* lines or just *modified* lines?

[This comment](https://github.com/editorconfig/editorconfig/issues/60#issuecomment-7105866) suggests it's modified only. In any case that's an important thing to consider when deciding whether or not to enable that property.

**Comment by jedmao** [comment]
That would be great to clarify the language on that feature. I can tell you that the Visual Studio plugin trims all trailing whitespace of all lines in the documents, but other editors might handle it differently. I don't know.

Either way, it should probably be noted in the docs, as you say.

**Comment by steverice** [comment]
Indeed; it may also be helpful for plugin authors to clarify the intent so there isn't that sort of inconsistency among implementations.

**Comment by xuhdev** [comment]
All plugins I can remember trim all lines. Do you know any plugins which do differently?

**Comment by jedmao** [comment]
No, but I've only used the VS and Sublime ones.

**Comment by atwellpub** [comment]
Notepad++ doesn't have an option to trim lines on save. A new macro has to be created. Not sure what to do.

**Comment by jedmao** [comment]
@atwellpub if that's a feature that's important to you have you considered using an editor that has support for that setting? I know that SublimeText has it and I'm sure others.

**Comment by atwellpub** [comment]
Sublime text does provide this feature already, but being a PHP developer I am very hapily dependent on Notepad++'s quick-nav function parser, which I can't find for sublime anywhere. 

![](http://content.screencast.com/users/adbox/folders/Jing/media/774c037d-77c1-4d26-b540-cc8f3bf9354b/2015-04-27_1254.png)

**Comment by jedmao** [comment]
@atwellpub by no means am I telling you what to use, but have you considered [PHPStorm](http://www.jetbrains.com/phpstorm/?fromMenu)? I'm willing to bet it has that feature. The people at [JetBrains](http://www.jetbrains.com/) know what they are doing!

**Comment by atwellpub** [comment]
I have done a small investigation and have not found that feature. I will gladly switch to an editor that provides me with both, or I understood how to write plugins for. 

Hopefully someone using PHPStorm will chime in.

**Comment by xuhdev** [comment]
but ... writing a macro isn't that hard?

**Comment by denofevil** [comment]
PhpStorm can do that: https://www.jetbrains.com/phpstorm/help/viewing-structure-of-a-source-file.html (not only for php)

**Comment by jedmao** [comment]
@xuhdev I agree that writing a macro for Notepad++ [shouldn't be hard](http://stackoverflow.com/questions/362444/how-to-write-macro-for-notepad), but I don't have any experience doing that in Notepad++, so I don't have much to offer.

It does seem that `Alt+Shift+S` in Notepad++ [will trim trailing whitespace](http://sourceforge.net/p/notepad-plus/feature-requests/1550/), but remembering to do that isn't terribly convenient, is it? It should be automated and bound to the EditorConfig setting.

I still think considering other editors might be the best viable option for @atwellpub, because of the features he deems important. @atwellpub, I can tell you that I've extensively used all 3 of the editors mentioned (Notepad++, PHPStorm and SublimeText 2/3) as well as Visual Studio. The important thing is to use the tool you find makes yourself the most productive. Based on what you need, it sounds like PHPStorm might be the right tool for you.

**Comment by aweibell** [comment]
I definitely would prefer having the three options that IntelliJ IDEA natively provides for what they name "Strip trailing spaces on save":
* __Modified lines__
* All
* None

Trailing spaces should not appear in code, but when I modify others code, I don't want my commit to include lots of white space corrections. It makes it way harder to see in a quick diff what __real__ code changes I did.

**Comment by jedmao** [comment]
I agree that there would be value in a `modified` setting:

```ini
[*]
trim_trailing_whitespace = modified
```
We can still use `true` for all lines and your None scenario above would be the existing default.

**Comment by aweibell** [comment]
@jedmao Exactly!

**Comment by jaller94** [comment]
Issues in code editors:
- Atom: Atom/whitespace#8
- VS Code: Microsoft/vscode#13157
